engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/ReentrantLocking.scala

```diff
@@ -126,4 +132,57 @@ class ReentrantLocking(logger: TruffleLogger) extends Locking {
     now2
   }
 
+  /** @inheritdoc */
+  override def getUniqueTaskIndex(): Int = {
+    orderLock.lock()
+    try {
+      taskIndex.enqueue()
+    } finally {
+      orderLock.unlock()
+    }
+  }
+
+  /** @inheritdoc */
+  override def waitOnTaskIndex(index: Int): Unit = {
+    orderLock.lock()
+    try {
+      val current = taskIndex.getCurrent()
+      while (current != index) {
+        indexAdvancedCondition.await()
+      }
+    } finally {
+      orderLock.unlock()
+    }
+  }
```

Comment:
How do we handle the possibility of `InterruptedException` here?

Reply:
How do we guarantee the liveness of the `PartiallyOrderedAsynchronousCommand` queue?

What happens if I have 4 threads with task ids 1, 2, 3 and 4. 1 is executing a long running operation, thus 2, 3 and 4 are waiting.

In the meantime, something kills the thread associated with (3). Once 1 finishes execution, 2 starts and then finishes. Now the `indexAdvancedCondtion` is signalled. But since (3) is dead, only (4) wakes up - it sees that its not yet its time - the current index is set at 3, and goes to sleep again. But now it will never wake up, because 3 will never advance the clock.